src/EventStore.Core/Services/VNode/ClusterVNodeController.cs

```diff
@@ -890,6 +879,26 @@ public class ClusterVNodeController : IHandle<Message> {
 
 		private void HandleAsReadOnlyReplica(GossipMessage.GossipUpdated message) {
 			if (_leader == null) throw new Exception("_leader == null");
+
+			var aliveLeaders = message.ClusterInfo.Members.Where(x => x.IsAlive && x.State == VNodeState.Leader);
+			var leaderCount = aliveLeaders.Count();
+
+			if (leaderCount == 0) {
+				Log.Debug(
+					"NO LEADER found in READ ONLY PRE-REPLICA/READ ONLY REPLICA state. Proceeding to READ ONLY LEADERLESS STATE. CURRENT LEADER: [{leader}]", _leader);
+				_stateCorrelationId = Guid.NewGuid();
+				_fsm.Handle(new SystemMessage.BecomeReadOnlyLeaderless(_stateCorrelationId));
+			}
+			else if (leaderCount == 1) {
+				var newLeader = VNodeInfoHelper.FromMemberInfo(aliveLeaders.First());
+				if (_leader.InstanceId != newLeader.InstanceId) {
+					Log.Information("LEADER CHANGE detected in READ ONLY PRE-REPLICA/READ ONLY REPLICA state. Proceeding to READ ONLY LEADERLESS STATE. CURRENT LEADER: [{leader}]. NEW LEADER: [{newLeader}].", _leader, newLeader);
+					_stateCorrelationId = Guid.NewGuid();
+					_fsm.Handle(new SystemMessage.BecomeReadOnlyLeaderless(_stateCorrelationId));
+				}
+			}
+			//if multiple leaders are found we just wait for the cluster to stabilize with a single leader during the next gossip updates
```

Comment:
Should we not move into a `ReadOnlyLeaderless` state if there are multiple leaders?

Reply:
@hayley-jean I've reasoned about it this way:
- If there is more than one leader, exactly zero or one of these leaders will have the quorum.
- If this node is not connected to the leader that has the quorum, then everything is fine, we do not need to transition to any state.
- If the node is connected to a leader that does not have the quorum, then the leader will eventually see that it does not have a quorum within the next few seconds and it will disconnect all slaves from it and go to `Unknown` state. The cluster will eventually stabilize with one leader and the read-only replica will transition to `ReadOnlyLeaderLess`

There is a case which is not considered in the above code though, if there are multiple leaders and the current leader is not in this list of leaders, we could definitely go directly to `ReadOnlyLeaderLess`. I will try to improve this part.

Reply:
@hayley-jean I've simplified the logic, also taking into consideration the above case in this commit: be04092b6b80792289597fc3634d37843454dbc2